### PR TITLE
 [AlibabaCloudLogService] sanitize labels for metrics

### PR DESCRIPTION
**Pull request by shabicheng**
**Description:** <Describe what has changed.>

OpenTelemetry metric's labels contains '.' or others none Letter or Digit characters, the alibabacloudlogserviceexporter do not replace these characters, so promQL will cause errors like "promql error: 1:33: parse error: unexpected character inside braces: '.'"

So we sanitize labels before send to sls, the code is almost copied from : https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/prometheusexporter/sanitize.go

**Link to tracking Issue:** <Issue number if applicable>

#3429 

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Review by qiansheng91 — COMMENTED** [review]
The prometheus exporter sanitizes the [metric name](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/prometheusexporter/collector.go#L80-L85) and [label key](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/prometheusexporter/collector.go#L92) and that the AlibabaCloudLogService export only sanitizes the label key,  Does the exporter need to concern it too?

**Comment by shabicheng** [comment]
> The prometheus exporter sanitizes the [metric name](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/prometheusexporter/collector.go#L80-L85) and [label key](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/prometheusexporter/collector.go#L92) and that the AlibabaCloudLogService export only sanitizes the label key, Does the exporter need to concern it too?

Metric names have already  sanitized in formatMetricName : https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go#L139

**Review by qiansheng91 — APPROVED** [review]
LGTM

File: exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go [hunk 1]
```diff
@@ -51,7 +51,14 @@ func (kv *KeyValues) Swap(i, j int) {
 	kv.keyValues[i], kv.keyValues[j] = kv.keyValues[j], kv.keyValues[i]
 }
 func (kv *KeyValues) Less(i, j int) bool { return kv.keyValues[i].Key < kv.keyValues[j].Key }
-func (kv *KeyValues) Sort()              { sort.Sort(kv) }
+
+func (kv *KeyValues) Sort() {
+	// sanitize label keys before sort
```
Comment (anuraaga) [inline exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go:56]: Shouldn't sanitization happen on insertion, not sorting? It seems mysterious for a sort method to modify the data itself.

Comment (shabicheng) [reply exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go]: done, sanitize keys when insert

File: exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go [hunk 2]
```diff
@@ -64,6 +68,7 @@ func (kv *KeyValues) Replace(key, value string) {
 
 func (kv *KeyValues) AppendMap(mapVal map[string]string) {
 	for key, value := range mapVal {
+		key = sanitize(key)
```
Comment (anuraaga) [inline exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go:71]: It looks like you can replace these 5 lines with calling the `Append` function

File: exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go [hunk 3]
```diff
@@ -51,9 +51,13 @@ func (kv *KeyValues) Swap(i, j int) {
 	kv.keyValues[i], kv.keyValues[j] = kv.keyValues[j], kv.keyValues[i]
 }
 func (kv *KeyValues) Less(i, j int) bool { return kv.keyValues[i].Key < kv.keyValues[j].Key }
-func (kv *KeyValues) Sort()              { sort.Sort(kv) }
+
+func (kv *KeyValues) Sort() {
```
Comment (anuraaga) [inline exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go:55]: Can we add unit tests for this file to demonstrate the sanitization?

File: exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go (hunk 2 above)
Comment (shabicheng) [reply exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go]: Yes, AppendMap can call Apend directly

File: exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go (hunk 3 above)
Comment (shabicheng) [reply exporter/alibabacloudlogserviceexporter/metricsdata_to_logservice.go:55]: Done

**Review by anuraaga — APPROVED** [review]

**Pull request merged**